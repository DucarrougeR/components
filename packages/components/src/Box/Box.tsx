import styled, { css } from 'styled-components'
import { flexbox, FlexboxProps } from 'styled-system'
import { CompatibleHTMLProps } from '@looker/design-tokens'

import { boxBaseCSS, BoxBaseProps } from './BoxBase'
// import { animation, AnimationProps } from './animation'
import { psuedoClasses, PsuedoProps } from './psuedoStyles'

const boxCSS = css`
  ${boxBaseCSS}
  ${flexbox}
  ${psuedoClasses}
`
export interface BoxProps
  extends CompatibleHTMLProps<HTMLDivElement>,
    BoxBaseProps,
    FlexboxProps,
    PsuedoProps {}

export const Box = styled.div<BoxProps>`
  ${boxCSS}
`
