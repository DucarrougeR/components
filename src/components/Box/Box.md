```js noeditor
const StatusAndResources = require('../../../styleguide_components/StatusAndResources').StatusAndResources;
<StatusAndResources
  status="experimental"
  githubURL="https://github.com/looker/relens/blob/master/src/components/Box/Box.tsx"
  feedbackTitle="Box Component Feedback" />
```

### About Box

`<Box />` is a low level component whose primary purpose is to provide a utility for easily controlling the spacing around elements in a consistent way using the [Lens spacing values](/#!/Spacing).

***

### Spacing with Box

The `<Box />` component allows you to control it's margin and padding properties for a chosen side.

Using spacing on a `<Box />` is like following a recipe, you select:
```js noeditor
<List type="number">
  <ListItem>The type of spacing, margin or padding</ListItem>
  <ListItem>The side to apply it to </ListItem>
  <ListItem>The amount of space to apply</ListItem>
</List>
```
```js noeditor
const BoxSpacingRecipeTableRender = require('../../../styleguide_components/BoxSpacingRecipeTable').BoxSpacingRecipeTableRender;
<Box my="xxlarge">
  <BoxSpacingRecipeTableRender />
</Box>
```
To get the correct spacing you choose one value from each of the above columns. For example, to get 8px of margin on the right side you would do. `<Box mr="small">` or if you wanted 40px of padding on all sides you would do `<Box p="xxlarge">`.

You can combine different spacing recipes to apply spacing to multiple sides at one time. For example if you wanted 16px of margin on the left and 20px of padding on the top you would do `<Box ml="medium" pt="large">`

#### Spacing Examples
Below are a few more examples of building a spacing recipe to help clarify how to use spacing with the `<Box>` component.
```js
  <Box m="large">Margin All Sides - Large</Box>
  <Box pl="xsmall">Padding left - Extra Small</Box>
  <Box pl="medium">Padding left -  Medium</Box>
  <Box my="xxxlarge" ml="xlarge">Margin Top & Bottom - XXXLarge, Margin left - XLarge</Box>
  <Box pb="xxlarge" ml="xxxlarge"> Padding Bottom - XXLarge, Margin left - XXXLarge</Box>
```

#### Spacing Responsively
Any one of the spacing properties can be passed an array of spacing sizes that will then apply at different breakpoints. There are 5 [breakpoints](/#!/Breakpoints) slots that you can target. Here is an example of what that looks like.
```js
  <Box pl={[
          'xsmall',   // xsmall spacing up to the first breakpoint
          'medium',   // medium spacing after first breakpoint
          'xlarge',   // xlarge spacing after second breakpoint
          'xxxlarge'  // xxxlarge spacing after third breakpoint
         ]}>
      My padding on the left changes with breakpoints
  </Box>
```
