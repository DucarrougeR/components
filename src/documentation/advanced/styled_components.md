[Styled Components](https://www.styled-components.com/) is a low level CSS
library used by Lens.

> ðŸ‘‹ Heads up! ðŸ‘‹
>
> **You should rarely need to use Styled Components.** Instead use the [`<Box/>`](/#/Components/Layout?id=box)
> component and its props which ensures you use Lens' theme values and
> stay within the Lens style guidelines. Styled Components should only be
> used for configuring [pseudo classes (eg `:hover`)](https://developer.mozilla.org/en-US/docs/Web/CSS/Pseudo-classes), [pseudo elements (eg
> `:before`)](https://developer.mozilla.org/en-US/docs/Web/CSS/Pseudo-elements), animations, or rarely when `<Box/>`'s properties are insufficient.

### Things You'll Learn

- Use Styled Components' `styled` function for styling pseudo classes and pseudo
  elements.
- Strip props out before passing them to lower level components to avoid runtime
  errors.
- Styled Components provides access to a component's prop values so styling can
  be computed dynamically.

## Styling a Lens Component

Let's pretend we want to build a new kind of `<Link>` that renders visited links
a bit differently. Our designer has laid out the following specs:

1.  Visited links should render red.
2.  The component should accept a prop called `visited` that renders the text as
    strike-through on hover.

To handle this, we'll use Styled Components' `styled` function because we'll
need to style both the `:visited` and `:hover` pseudo classes. Step one is to
set the `:visited` color to our palette.red400 color. We can access the Theme
via Styled Components, so this is easy enough:

```js
import styled from 'styled-components'
import { Link } from '../../components/Link'

const VisitedLink = styled(Link)`
  :visited {
    color: ${props => props.theme.colors.palette.red400};
  }
`
;<VisitedLink href="https://www.google.com">Hello</VisitedLink>
```

Note, `styled` expects to be passed a React component directly. In the example
above, we pass it the `Link` Lens component.

Now let's handle the `visited` prop. Things become a bit more complex because we
need to add a new prop to our Link. In Typescript you might expect to do
something like this, but it will cause a run time error like `Warning: Received 'true' for a non-boolean attribute 'visited'`:

```ts noeditor
import { Link } from '../../components/Link'

interface VisitedLink extends Link {
  visited?: boolean
}

const VisitedLink = styled(Link)`
  :visited {
    color: ${props => props.theme.colors.palette.red400};
  }

  :hover {
    text-decoration: ${props => (props.visited ? 'line-through' : 'underline')};
  }
`
;<VisitedLink href="https://www.google.com">Hello</VisitedLink>
```

This happens because when Styled Components wraps a component, it passes all
props passed to it straight through to the wrapped component (`<Link>` in this
case). While this may seem confusing or inconvenient, [this behavior is by
design](https://www.styled-components.com/docs/basics#passed-props).

To eliminate the runtime warning, we need to intercept the prop and prevent it
from being passed to `<Link/>`. It requires a bit of ceremony, but is easy
enough with [ES6's spread syntax](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Operators/Spread_syntax):

```js
import styled from 'styled-components'
import { Link } from '../../components/Link'

const VisitedLinkFactory = ({ visited, ...props }) => <Link {...props} />

const VisitedLink = styled(VisitedLinkFactory)`
  :visited {
    color: ${props => props.theme.colors.palette.red400};
  }

  :hover {
    text-decoration: ${props => (props.visited ? 'line-through' : 'underline')};
  }
`
;<VisitedLink visited href="https://www.google.com">
  Hello
</VisitedLink>
```

Et voilÃ . Note we created the internal `VisitedLinkFactory` function
which strips out the `visited` prop before it ever reaches the composed `Link`
component.

## Styling Child Components and Linter Errors

Styled Components [provides a useful mechanism for styling child components](https://www.styled-components.com/docs/advanced#referring-to-other-components). This syntax, unfortunately, [generates a known linter error](https://github.com/styled-components/stylelint-processor-styled-components/issues/34#issuecomment-325339118). To remove the linter error, simply add the `/* sc-selector */` tag to your style definition. For example:

```jsx static
import { styled, Link } from 'looker-lens'

const Card = styled.div`
  ${/* sc-selector */ Link} {
    background-color: ${props => props.theme.palette.green300};
  }
`
```
