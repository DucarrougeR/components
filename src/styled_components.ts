// Required by ThemeProvider to compile the proper declaration files.
import * as React from 'react'
import * as styledComponents from 'styled-components'
// tslint:disable-next-line:no-duplicate-imports
import { ThemedStyledProps } from 'styled-components'
import { ThemeInterface } from './themes'
export { StyledComponentClass, Styles } from 'styled-components'

const { css, injectGlobal, keyframes, ThemeProvider } = styledComponents

// Typescript complains that the `as ThemedStyledComponentsModule` performs
// unnecessary casting, which is not true. Without this line the Themes
// attached to component prop types would not type-check correctly.
//
// tslint:disable-next-line:no-unnecessary-type-assertion
const styled = (styledComponents as styledComponents.ThemedStyledComponentsModule<
  ThemeInterface
>).default

export { css, injectGlobal, keyframes, React, ThemeProvider, ThemedStyledProps }
export default styled
