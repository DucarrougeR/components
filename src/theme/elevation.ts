export type ElevationLayers = '1' | '2' | '3' | '4' | '5' | '6' | '7' | 'max'

export type Shadows = { [K in ElevationLayers]?: string }

export const shadows: Shadows = {
  1: '0 1px 2px rgba(0, 0, 0, 0.1), 0px 1px 1px rgba(0, 0, 0, 0.05)',
  2: '0 3px 4px rgba(0, 0, 0, 0.11), 0 1px 2px rgba(0, 0, 0, 0.04)',
  3: '0 6px 10px rgba(0, 0, 0, 0.12), 0 1px 2px rgba(0, 0, 0, 0.04)',
  4: '0 9px 18px rgba(0, 0, 0, 0.14), 0 1px 2px rgba(0, 0, 0, 0.04),  0 10px 8px 4px rgba(0, 0, 0, 0.01)',
  5: '0 14px 20px rgba(0, 0, 0,.15), 0 1px 1px rgba(0, 0, 0, 0.04),  0 14px 10px 8px rgba(0, 0, 0, 0.02)',
  6: '0 22px 42px 0px rgba(0, 0, 0,.18),  0 1px 1px rgba(0, 0, 0, 0.05), 0 16px 20px 10px rgba(0, 0, 0, 0.025)',
}
