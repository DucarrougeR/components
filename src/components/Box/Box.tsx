import tag from 'clean-tag'
import * as React from 'react'
import {
  alignContent,
  AlignContentProps,
  alignItems,
  AlignItemsProps,
  alignSelf,
  AlignSelfProps,
  border,
  borderBottom,
  BorderBottomProps,
  borderColor,
  BorderColorProps,
  borderLeft,
  BorderLeftProps,
  BorderProps,
  borderRadius,
  BorderRadiusProps,
  borderRight,
  BorderRightProps,
  borderTop,
  BorderTopProps,
  bottom,
  BottomProps,
  boxShadow,
  BoxShadowProps,
  color,
  ColorProps,
  display,
  DisplayProps,
  flex,
  flexBasis,
  FlexBasisProps,
  flexDirection,
  FlexDirectionProps,
  FlexProps,
  flexWrap,
  FlexWrapProps,
  fontFamily,
  FontFamilyProps,
  fontSize,
  FontSizeProps,
  fontStyle,
  FontStyleProps,
  fontWeight,
  FontWeightProps,
  height,
  HeightProps,
  justifyContent,
  JustifyContentProps,
  left,
  LeftProps,
  letterSpacing,
  LetterSpacingProps,
  lineHeight,
  LineHeightProps,
  maxHeight,
  MaxHeightProps,
  maxWidth,
  MaxWidthProps,
  minHeight,
  MinHeightProps,
  minWidth,
  MinWidthProps,
  opacity,
  OpacityProps,
  order,
  OrderProps,
  overflow,
  OverflowProps,
  position,
  PositionProps,
  right,
  RightProps,
  space,
  textAlign,
  TextAlignProps,
  top,
  TopProps,
  verticalAlign,
  VerticalAlignProps,
  width,
  WidthProps,
  zIndex,
  ZIndexProps,
} from 'styled-system'
import { css, LensSpaceProps, reset, styled } from '../../style'
import { Omit } from '../../types'
import { cursor, CursorProps } from './style_utilities'

const Tag = tag

export interface BoxFlexProps
  extends AlignContentProps,
    AlignItemsProps,
    FlexDirectionProps,
    FlexWrapProps,
    JustifyContentProps {}

export interface BoxFlexItemProps
  extends AlignSelfProps,
    FlexBasisProps,
    FlexProps,
    OrderProps {}

export type StyledSystemCompatibleHTMLProps<T> = Omit<
  React.HTMLProps<T>,
  'width' | 'color' | 'height' | 'is' | 'size'
>

export interface BoxBaseProps<T>
  extends StyledSystemCompatibleHTMLProps<T>,
    LensSpaceProps,
    BorderColorProps,
    BorderProps,
    BorderRadiusProps,
    BorderRightProps,
    BorderLeftProps,
    BorderBottomProps,
    BorderTopProps,
    BottomProps,
    BoxShadowProps,
    ColorProps,
    CursorProps,
    DisplayProps,
    FontFamilyProps,
    FontSizeProps,
    FontStyleProps,
    FontWeightProps,
    HeightProps,
    LeftProps,
    LetterSpacingProps,
    LineHeightProps,
    MaxHeightProps,
    MaxWidthProps,
    MinHeightProps,
    MinWidthProps,
    OpacityProps,
    OverflowProps,
    PositionProps,
    RightProps,
    TextAlignProps,
    TopProps,
    VerticalAlignProps,
    WidthProps,
    ZIndexProps {
  className?: string
  is?: string | React.ReactNode
  ref?: React.Ref<any>
  style?: React.CSSProperties
}

export interface BoxProps<T>
  extends BoxBaseProps<T>,
    BoxFlexProps,
    BoxFlexItemProps {}

export type BoxBasePropsWithout<T, Keys> = Omit<BoxBaseProps<T>, Keys>
export type BoxPropsWithout<T, Keys> = Omit<BoxProps<T>, Keys>

const cursorPointerOnClick = (props: BoxProps<HTMLElement>) =>
  props.onClick &&
  !props.disabled &&
  css`
    cursor: pointer;
  `

export const Box = styled<BoxProps<HTMLElement>>(Tag)`
  /**
   * Global reset applied to prevent styling on top level tags outside of Lens
   * from interfering with Lens styles.
   *
   * This **must** be first.
   */
  ${reset};

  /**
   * Rules here should provide convenience styling for Box derived components.
   * Generally anything here could be overwritten by explicit values set via
   * Box's prop values. For example a function here that sets 'cursor: pointer'
   * would be overwritten by an explicit <Box cursor='copy'/>.
   */
  ${cursorPointerOnClick};

  /**
   * Style Utilities that extend Box's props. Most of these come from
   * styled-system but some are Lens specific.
   *
   * These should be last to override rules with lower priority.
   */
  ${alignContent};
  ${alignItems};
  ${alignSelf};
  ${border};
  ${borderRadius};
  ${borderRight};
  ${borderLeft};
  ${borderBottom};
  ${borderTop};
  ${borderColor};
  ${bottom};
  ${boxShadow};
  ${color};
  ${cursor};
  ${display};
  ${flex};
  ${flexBasis};
  ${flexDirection};
  ${flexWrap};
  ${fontFamily};
  ${fontSize};
  ${fontStyle};
  ${fontWeight};
  ${height};
  ${justifyContent};
  ${left};
  ${letterSpacing};
  ${lineHeight};
  ${maxHeight};
  ${maxWidth};
  ${minHeight};
  ${minWidth};
  ${opacity};
  ${order};
  ${overflow};
  ${position};
  ${right};
  ${space};
  ${textAlign};
  ${top};
  ${verticalAlign};
  ${width};
  ${zIndex};
`
