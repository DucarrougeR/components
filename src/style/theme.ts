import { rem } from 'polished'
import { InterpolationValue } from 'styled-components'
import {
  SizeLarge,
  SizeMedium,
  SizeNone,
  SizeSmall,
  SizeXLarge,
  SizeXSmall,
  SizeXXLarge,
  SizeXXXLarge,
} from '../types'
import { Easings, easings } from './easings'
import { shadows, Shadows } from './elevation'
import { fontFaces, FontFaces } from './font_faces'
import { FontRamp, fontSizes, lineHeights } from './font_sizes'
import { fontWeights } from './font_weights'
import { palette, Palette } from './palette'
import { SemanticColors, semanticColors } from './semantic_colors'
import { transitions, Transitions } from './transitions'

export * from './easings'
export * from './elevation'
export * from './font_sizes'
export * from './palette'
export * from './semantic_colors'
export * from './transitions'

export type SpacingSizes =
  | SizeNone
  | SizeLarge
  | SizeMedium
  | SizeSmall
  | SizeXLarge
  | SizeXSmall
  | SizeXXLarge
  | SizeXXXLarge

export interface Theme {
  breakpoints: string[]
  easings: Easings
  fontFaces: FontFaces
  fontSizes: FontRamp
  fontWeights: Record<string, number>
  lineHeights: FontRamp
  palette: Palette
  /**
   * A function that can be overridden to return different reset css properties
   * or null to remove all resets. Most base elements in Lens implement the reset.
   */
  reset: () => InterpolationValue
  semanticColors: SemanticColors
  shadows: Shadows
  space: Record<SpacingSizes, string>
  transitions: Transitions
}

export const theme: Theme = {
  breakpoints: ['30rem', '48rem', '64rem', '75rem', '90rem'],
  easings,
  fontFaces,
  fontSizes,
  fontWeights,
  lineHeights,
  palette,
  reset: () => {
    return {
      border: 0,
      font: 'inherit',
      fontSize: '100%',
      margin: 0,
      padding: 0,
      verticalAlign: 'baseline',
    }
  },
  semanticColors,
  shadows,
  space: {
    // tslint:disable:object-literal-sort-keys
    none: 'none',
    xsmall: rem(4),
    small: rem(8),
    medium: rem(16),
    large: rem(20),
    xlarge: rem(32),
    xxlarge: rem(40),
    xxxlarge: rem(60),
    // tslint:enable:object-literal-sort-keys
  },
  transitions,
}
