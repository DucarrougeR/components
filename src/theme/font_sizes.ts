import { rem } from 'polished'
import { RampSizes } from '../style/ramp_sizes'

export type FontRamp = { [K in RampSizes]: string }

// tslint:disable:object-literal-sort-keys
export const fontRamp: FontRamp = {
  d1: rem(58),
  d2: rem(46),
  d3: rem(36),
  1: rem(25),
  2: rem(22),
  3: rem(18),
  4: rem(16),
  5: rem(14),
  6: rem(12),
}

// tslint:disable:object-literal-sort-keys
export const lineHeightRamp: FontRamp = {
  d1: rem(84),
  d2: rem(64),
  d3: rem(52),
  1: rem(40),
  2: rem(32),
  3: rem(28),
  4: rem(24),
  5: rem(20),
  6: rem(16),
}
