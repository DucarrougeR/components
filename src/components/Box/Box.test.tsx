import 'jest-styled-components'
import * as React from 'react'
import { create } from 'react-test-renderer'
import { ThemeProvider } from '../../styled_components'
import theme from '../../themes'
import { Box } from './Box'

test('Box default', () => {
  const component = create(<Box theme={theme}>ðŸ¥‘</Box>)
  const tree = component.toJSON()
  expect(tree).toMatchSnapshot()
})
