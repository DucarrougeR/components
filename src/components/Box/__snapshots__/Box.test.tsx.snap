// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Box default 1`] = `
<div
  className=""
>
  ðŸ¥‘
</div>
`;
